src/common/download.cpp

```diff
@@ -70,69 +120,145 @@ namespace tools
+        boost::lock_guard<boost::mutex> lock(control->mutex);
+        MERROR("Failed invoking GET command to " << control->uri << ", no status info returned");
         client.disconnect();
-        return false;
+        control->result_cb(control->path, control->uri, control->success);
+        return;
       }
       MDEBUG("response code: " << info->m_response_code);
+      MDEBUG("response length: " << info->m_header_info.m_content_length);
       MDEBUG("response comment: " << info->m_response_comment);
       MDEBUG("response body: " << info->m_body);
       for (const auto &f: info->m_additional_fields)
         MDEBUG("additional field: " << f.first << ": " << f.second);
       if (info->m_response_code != 200)
       {
+        boost::lock_guard<boost::mutex> lock(control->mutex);
         MERROR("Status code " << info->m_response_code);
         client.disconnect();
-        return false;
+        control->result_cb(control->path, control->uri, control->success);
+        return;
       }
       client.disconnect();
       f.close();
       MDEBUG("Download complete");
-      return true;
+      lock.lock();
+      control->success = true;
+      control->result_cb(control->path, control->uri, control->success);
+      return;
     }
     catch (const std::exception &e)
     {
       MERROR("Exception in download thread: " << e.what());
-      return false;
+      // fall through and call result_cb not from the catch block to avoid another exception
     }
+    boost::lock_guard<boost::mutex> lock(control->mutex);
+    control->result_cb(control->path, control->uri, control->success);
   }
 
-  bool download(const std::string &path, const std::string &url)
+  bool download(const std::string &path, const std::string &url, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> cb)
   {
-    bool success;
-    std::unique_ptr<boost::thread> thread(new boost::thread([&](){ success = download_thread(path, url); }));
-    thread->join();
+    bool success = false;
+    download_async_handle handle = download_async(path, url, [&success](const std::string&, const std::string&, bool result) {success = result;}, cb);
+    download_wait(handle);
     return success;
   }
+
+  download_async_handle download_async(const std::string &path, const std::string &url, std::function<void(const std::string&, const std::string&, bool)> result, std::function<bool(const std::string&, const std::string&, size_t, ssize_t)> progress)
+  {
+    download_async_handle control = std::make_shared<download_thread_control>(path, url, result, progress);
+    control->thread = boost::thread([control](){ download_thread(control); });
+    return control;
+  }
+
+  bool download_finished(const download_async_handle &control)
+  {
+    CHECK_AND_ASSERT_MES(control != 0, false, "NULL async download handle");
+    boost::lock_guard<boost::mutex> lock(control->mutex);
+    return !control->thread.joinable();
```

Comment:
And this doesn't work anymore. `joinable()` remains true until someone detaches or joins. Since I told you to move the detach call from the `download_thread` function, this will only be true if someone calls `download_error` or `download_cancel` first.

So bad idea.